### ISSUE TITLE
attention inference slow

### ISSUE THREAD

**Issue by chenzhanyiczy** [HazyResearch/flash-attention#262]
The speed of flash-attention(QKV/softmax fused part) in megatron-LM  is much slower than the native implementation of megatron-LM.  What is the reason?  
I used the latest version：https://github.com/NVIDIA/megatron-lm
I suspect that q, k, and v need a large number of copies caused by contiguous. Is this ‘contiguous’ necessary?
`q, k, v = [rearrange(x, 's b ... -> b s ...').contiguous()
                       for x in (query_layer, key_layer, value_layer)]
context_layer = self.core_attention_flash(q, k, v)
context_layer = rearrange(context_layer, 'b s h d -> s b (h d)').contiguous()`

**Comment by tridao** [comment]
You can try removing the `contiguous()`.
And you can benchmark different implementations with the shape that you're gonna use.

Without knowing the input shapes and GPU type, it's hard to say why something is fast or slow.

In general FlashAttention is optimized for training. For inference it's often used for prompt processing. I don't think it's helpful for iterative decoding where q only has 1 token.

**Comment by chenzhanyiczy** [comment]
> 

thanks,. I removed the contiguous(), but the result is wrong, why is this?
Env:
model: llama-13B(layer:40, hiden_size:5120, head_size:40)
cuda:11.8
pytorch:1.13
megatron-LM:newest

**Comment by tridao** [comment]
FlashAttention expects input format (batch, seqlen, ...) while megatron uses the format (seqlen, batch, ...).
And inputs should be contiguous.

Without careful benchmarking, it's hard to say whether it is slow or why.

As mentioned if you're using Megatron for iterative decoding (not for training or prompt-processing) I don't recommend using FlashAttention.

**Comment by chenzhanyiczy** [comment]
> FlashAttention expects input format (batch, seqlen, ...) while megatron uses the format (seqlen, batch, ...). And inputs should be contiguous.
> 
> Without careful benchmarking, it's hard to say whether it is slow or why.
> 
> As mentioned if you're using Megatron for iterative decoding (not for training or prompt-processing) I don't recommend using FlashAttention.

Does input's memory must be contiguous?

**Comment by tridao** [comment]
I think so.
Strictly speaking it only needs the last dimension to be contiguous (i.e. with stride 1). Calling `.contiguous()` is a quick way of achieving that.
